lib/host_runner.rb

```diff
@@ -35,10 +47,60 @@ def start
       raise LockError.new("Host '#{@host_name}' already locked")
     end
 
+    connect
     @ip
   end
 
+  def cleanup
+    run_command "snapper", "create", "-c", "number", "--pre-number", @base_snapshot_id.to_s,
+      "--description", "pennyworth_snapshot"
+    run_command "snapper", "undochange", "#{@base_snapshot_id}..0"
+    run_command "bash", "-c", "reboot &"
+  end
+
   def stop
+    if @connected && !ENV["SKIP_CLEANUP"]
+      cleanup
+    end
     @locker.release_lock(@host_name)
   end
+
+  private
+
+  # Makes sure the we can connect to the remote system as root (without a
+  # password or passphrase) and that all required binaries are available.
+  def connect
+    begin
+      Cheetah.run "ssh", "-q", "-o", "BatchMode=yes", "root@#{@ip}"
+    rescue Cheetah::ExecutionFailed
+      raise SshConnectionFailed.new(
+        "Could not establish SSH connection to host '#{@ip}'. Please make sure that " \
+        "you can connect non-interactively as root, e.g. using ssh-agent.\n\n" \
+        "To copy your default ssh key to the machine run:\n" \
+        "ssh-copy-id root@#{@ip}"
```

Comment:
Since `pennyworth` can't depend on `machinery` that would mean that the `RemoteSystem` depends on code from `pennyworth` and that doesn't feel right to me.